jetty9-compat-base/src/main/java/com/google/apphosting/runtime/jetty9/NoOpSession.java

```diff
@@ -0,0 +1,186 @@
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.ServletContext;
+import javax.servlet.http.HttpSession;
+import javax.servlet.http.HttpSessionContext;
+
+
+/**
+ * A minimal implementation of HttpSession, just sufficient to keep and id attributes within a
+ * single request.
+ */
+public class NoOpSession implements HttpSession {
+
+  protected ServletContext _context;
+  protected String _clusterId;
+  protected String _nodeId;
+  protected boolean _valid = true;
+  private final Map<String, Object> _attributes = new HashMap<String, Object>();
+
+  /**
+   * @param id
+   * @param nodeId
+   * @param context
+   */
+  protected NoOpSession(String id, String nodeId, ServletContext context) {
+    _clusterId = id;
+    _nodeId = nodeId;
+    _context = context;
+  }
+
+  /**
+   * Finish a request
+   */
+  public void complete() {
+    _attributes.clear();
+  }
+
+
+  /**
+   * @see javax.servlet.http.HttpSession#getCreationTime()
+   */
+  @Override
+  public long getCreationTime() {
+    return 0;
+  }
+
+
+  @Override
+  public String getId() {
+    return _clusterId;
+  }
+
+  /**
+   * @return the session id with worker suffix
+   */
+  public String getNodeId() {
+    return _nodeId;
+  }
+
+
+  @Override
+  public long getLastAccessedTime() {
+    return 0;
+  }
+
+
+  @Override
+  public ServletContext getServletContext() {
+    return _context;
+  }
+
+
+  @Override
+  public void setMaxInactiveInterval(int interval) {
+  }
+
+
+  @Override
+  public int getMaxInactiveInterval() {
+    return 0;
+  }
+
+
+  @Override
+  public HttpSessionContext getSessionContext() {
+    return null;
+  }
+
+
+  @Override
+  public Object getAttribute(String name) {
+    return _attributes.get(name);
+  }
+
+
+  @Override
+  public Object getValue(String name) {
+    return getAttribute(name);
+  }
+
+
+  @Override
+  public Enumeration<String> getAttributeNames() {
+    List<String> names = _attributes == null
+      ? Collections.EMPTY_LIST
+      : new ArrayList<String>(_attributes.keySet());
+    return Collections.enumeration(names);
+  }
+
+
+  @Override
+  public String[] getValueNames() {
+    List<String> names = _attributes == null
+      ? Collections.EMPTY_LIST
+      : new ArrayList<String>(_attributes.keySet());
+    if (names.isEmpty()) {
+      return null;
+    }
+    return names.toArray(new String[names.size()]);
+  }
+
+
+  @Override
+  public void setAttribute(String name, Object value) {
+    _attributes.put(name, value);
+  }
+
+
+  @Override
+  public void putValue(String name, Object value) {
+    setAttribute(name, value);
+  }
+
+
+  @Override
+  public void removeAttribute(String name) {
+    _attributes.remove(name);
+  }
+
+
+  @Override
+  public void removeValue(String name) {
+    removeAttribute(name);
+  }
+
+
+  @Override
+  public void invalidate() {
+    _valid = false;
```

Comment:
should all methods, including this one, have a `Preconditions.checkState(_valid, "Session is  not valid");` ?